common/src/main/java/com/gregtechceu/gtceu/api/data/worldgen/generator/indicators/SurfaceIndicatorGenerator.java

```diff
@@ -0,0 +1,187 @@
+import com.gregtechceu.gtceu.api.data.worldgen.ores.GeneratedVeinMetadata;
+import com.gregtechceu.gtceu.api.data.worldgen.ores.OreIndicatorPlacer;
+import com.gregtechceu.gtceu.common.block.SurfaceRockBlock;
+import com.gregtechceu.gtceu.common.data.GTBlocks;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import com.tterrag.registrate.util.nullness.NonNullSupplier;
+import lombok.AllArgsConstructor;
+import net.minecraft.MethodsReturnNonnullByDefault;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.core.SectionPos;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.util.RandomSource;
+import net.minecraft.util.StringRepresentable;
+import net.minecraft.util.valueproviders.ConstantFloat;
+import net.minecraft.util.valueproviders.ConstantInt;
+import net.minecraft.util.valueproviders.FloatProvider;
+import net.minecraft.util.valueproviders.IntProvider;
+import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.WorldGenLevel;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.chunk.BulkSectionAccess;
+import net.minecraft.world.level.levelgen.Heightmap;
+import org.apache.commons.lang3.function.TriFunction;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+
+@MethodsReturnNonnullByDefault
+@ParametersAreNonnullByDefault
+public class SurfaceIndicatorGenerator extends IndicatorGenerator {
+    public static final Codec<SurfaceIndicatorGenerator> CODEC = RecordCodecBuilder.create(instance -> instance.group(
+            BuiltInRegistries.BLOCK.byNameCodec().fieldOf("block").forGetter(ext -> ext.block.get()),
+            IntProvider.codec(1, 32).fieldOf("radius").forGetter(ext -> ext.radius),
+            FloatProvider.codec(0.0f, 1.0f).fieldOf("density").forGetter(ext -> ext.density),
+            StringRepresentable.fromEnum(IndicatorPlacement::values).fieldOf("placement").forGetter(ext -> ext.placement)
+    ).apply(instance, SurfaceIndicatorGenerator::new));
+
+    private NonNullSupplier<? extends Block> block = NonNullSupplier.of(() -> Blocks.AIR);
+    private IntProvider radius = ConstantInt.of(5);
+    private FloatProvider density = ConstantFloat.of(0.2f);
+    private IndicatorPlacement placement = IndicatorPlacement.SURFACE;
+
+    @AllArgsConstructor
+    public enum IndicatorPlacement implements StringRepresentable {
+        SURFACE((level, access, pos) -> pos.atY(
+                level.getHeight(Heightmap.Types.OCEAN_FLOOR_WG, pos.getX(), pos.getZ())
+        ),  state -> state.setValue(SurfaceRockBlock.FACING, Direction.DOWN)),
```

Comment:
TRUSTING THAT THE BLOCK PROVIDED HAS THAT PROPERTY

Reply:
Right now this only allows surface rock indicators, which do have the property.  
But I might honestly just allow arbitrary blocks instead.

Reply:
well the codec allows any block

Reply:
Ah true, I only accounted for the method to set the type.